include/deal.II/base/memory_space.h

```diff
@@ -0,0 +1,203 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_memory_space_h
+#define dealii_memory_space_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/cuda.h>
+#include <deal.II/base/exceptions.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Structure describing Host memory space.
+ */
+struct Host
+{};
+
+
+
+/**
+ * Structure describing CUDA memory space.
+ */
+struct CUDA
+{};
```

Comment:
Would it make sense to encapsulate these two structs into a namespace `MemorySpaces`, or maybe also `MemorySpaceData`? I am a bit hesitant to import `Host` and `CUDA` as a top-level name. It would also fit well with the class `MemorySpace` you use as a generic template name.